073-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,339 @@
+4. Maybe deleting an old CA certificate if it is not longer used
+   1. This only occurs if all pods in the cluster have the correct strimzi.io/cluster-ca-cert-generation annotation, meaning all the component certificates have been signed by the latest Cluster CA certificate.
+
+ZooKeeperReconciler and KafkaReconciler classes:
+1. Generate the certificates the Kafka and ZooKeeper nodes will present to clients (other ZK/Kafka nodes, the Cluster Operator, Kafka producer and consumers etc) and store them in the Kubernetes Secrets listed above, with the cluster-ca-cert-generation (and clients-ca-cert-generation for Kafka) annotation set to the current value passed in with the Cluster CA (and Clients CA).
+2. Create/update the StrimziPodSet used for the nodes in the cluster and add the following annotations:
+   1. cluster-ca-cert-generation set to the current values passed in with the Cluster CA.
+   2. cluster-ca-key-generation set to the current values passed in with the Cluster CA.
+   3. server-cert-hash set to the full SHA1-hash of the certificate that was generated.
+
+EntityOperatorReconciler, CruiseControlReconciler and KafkaExporterReconciler generate the certificates they will use to authenticate with Kafka and ZooKeeper and store them in the Kubernetes Secrets listed above, with the cluster-ca-cert-generation set to the current value passed in with the Cluster CA.
+
+### Example flows
+
+When a new cluster is first created the flow is:
+
+1. Either the user or the CaReconciler creates the Cluster CA certificate and key and stores them in the <CLUSTER_NAME>-cluster-ca-cert and <CLUSTER_NAME>-cluster-ca Kubernetes Secrets respectively.
+2. The CaReconciler generates the Cluster Operator certificate and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret.
+3. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc) generate their certificates, store them in Kubernetes Secrets and annotate the Kubernetes Secrets and Pods with the cert and key generations, and (for ZooKeeper and Kafka) the certificate hash.
+
+When the Cluster CA private key is updated (either because it has expired or the user has incremented the generation to indicate they've manually updated the Kubernetes Secret):
+1. If Strimzi is managing the Cluster CA, the CaReconciler renews the Cluster CAs certificate and key. The old certificate and key are kept in the respective Kubernetes Secret and renamed to ca-YYYY-MM-DDTHH-MM-SSZ.crt.
+2. The CaReconciler skips updating the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret (since the components still only trust the old CA certificate).
+3. The CaReconciler rolls the ZooKeeper and Kafka nodes to trust the new CA certificates.
+4. The ZooKeeper and Kafka reconcilers:
+   1. Generate new certificates signed by the new CA key.
```

Comment:
Take into account that this is not the default behaviour in Strimzi.
We have the `clusterCA.certificateExpirationPolicy` field which defines if on certificate expiration, Strimzi should just renew the CA certificate (keeping the private key) using `renew-certificate` value or starting from a new private key using `replace-key` value. As I mentioned above regenerating a private key is not an operation which runs so often unless it's compromised, so we made the renewal (keeping the private key) as the default behaviour. More details here https://github.com/strimzi/strimzi-kafka-operator/blob/main/api/src/main/java/io/strimzi/api/kafka/model/common/CertificateExpirationPolicy.java
Said that, we should be clear about this in the example flow you are describing.